sql/sql_thd_api.cc

```diff
@@ -478,7 +478,21 @@ char *thd_security_context(MYSQL_THD thd, char *buffer, size_t length,
     else
       len = min(thd->query().length, max_query_len);
     str.append('\n');
-    str.append(thd->query().str, len);
+
+    /* The above limitation of length can cut the query string character
+       in the middle if it contains multbyte characters.
+       Let's put it through convert_string() method which will detect
+       such a case and produce valid string, truncating spoiled, trailing
+       character. */
+    LEX_STRING truncated_query = {nullptr, 0};
+    if (len < thd->query().length &&
+        !thd->convert_string(&truncated_query, thd->charset(), thd->query().str,
```

Comment:
It looks like `convert_string()` allocates memory that needs to be freed afterwards

Reply:
Yes, it allocates from `Query_arena`, so it is freed automatically after the query is done.

Reply:
Oh, I see now. Ignore then.